### PR TITLE
[ONNX] Add support for opset14 in PT-ONNX exporter

### PR DESCRIPTION
**Pull request by shubhambhokare1**
- Change onnxruntime version 1.7.0 -> 1.8.0 in CI
- Add symbolics for new operators (HardSwish, Trilu)
- Add symbolics for updated operators including BatchNorm, Reshape (with allowzero)
- Enable constant folding for opset 14
- Enable ORT tests for opset 14

### CODE REVIEW

File: test/onnx/test_utility_funs.py [hunk 1]
```diff
@@ -932,6 +937,10 @@ def forward(self, x, y):
                                            (TestCase,),
                                            dict(TestUtilityFuns.__dict__, opset_version=13))
 
+# opset 14 tests
+TestUtilityFuns_opset14_new_jit_API = type(str("TestUtilityFuns_opset14_new_jit_API"),
```
Comment (neginraoof) [inline test/onnx/test_utility_funs.py:941]: All these tests for new API are not useful anymore, since new API is now default.
I have an open PR to remove these tests. We can remove this line too.

Comment (shubhambhokare1) [reply test/onnx/test_utility_funs.py]: Noted, thanks :)

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 2]
```diff
@@ -5590,6 +5600,38 @@ def forward(self, x):
...
+    @skipIfUnsupportedMinOpsetVersion(14)
+    def test_tril(self):
+        class trilModel(torch.nn.Module):
+            def forward(self, x):
+                return torch.tril(x)
+
+        x = torch.randn(2, 3, 4)
+        self.run_test(trilModel(), (x))
+
+        class trilModelwithDiagonal(torch.nn.Module):
```
Comment (fatcat-z) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:5612]: Can we add a new test for negative value of Diagonal, like -1? The logic of this is different than positive one.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 3]
```diff
@@ -5590,6 +5600,38 @@ def forward(self, x):
...
+    @skipIfUnsupportedMinOpsetVersion(14)
+    def test_triu(self):
+        class triuModel(torch.nn.Module):
+            def forward(self, x):
+                return torch.triu(x)
+
+        x = torch.randn(2, 3, 4)
+        self.run_test(triuModel(), (x))
+
+        class triuModelwithDiagonal(torch.nn.Module):
```
Comment (fatcat-z) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:5628]: Can we add a new test for positive value of Diagonal, like 1?

File: torch/csrc/jit/passes/onnx/constant_fold.cpp [hunk 4]
```diff
@@ -260,7 +260,8 @@ c10::optional<at::Tensor> runTorchBackendForOnnx(
       return runTorchSlice_opset9(node, inputTensorValues);
     } else if (
         opset_version == ONNX_OPSET_10 || opset_version == ONNX_OPSET_11 ||
-        opset_version == ONNX_OPSET_12 || opset_version == ONNX_OPSET_13) {
+        opset_version == ONNX_OPSET_12 || opset_version == ONNX_OPSET_13 ||
+        opset_version == ONNX_OPSET_14) {
```
Comment (fatcat-z) [inline torch/csrc/jit/passes/onnx/constant_fold.cpp:264]: For each new ops version, seems like we always need to update 2 places: here and adding a new ops version, like ONNX_OPSET_xx, in constant_fold.h. How about optimizing the code so that we only need to update 1 file constant_fold.h in future:

1. Add a new static vector in constant.fold.h file. This vector will be initialized with all existing opsets, from ONNX_OPSET_11 to ONNX_OPSET_14.
2. In here, we just call find() method to check if opset_version exists in this vector, instead of having so many OR statements.

In future, if we have a new opset 15, we just need to add it into the static vector defined in constant_fold.h file. No more work needs to be done here.

Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/constant_fold.cpp:264]: good point. I also think this condition is equivalent to `opset_version >= ONNX_OPSET_10`.

File: torch/onnx/symbolic_opset14.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+def triu(g, self, diagonal, out=None):
+    k = g.op("Constant", value_t=torch.tensor(diagonal, dtype=torch.int64))
+    return g.op("Trilu", self, k, upper_i=1)
+
+@parse_args("v", "v")
+def reshape(g, self, shape):
+    shape = sym_help._maybe_get_const(shape, "is")
+    if not sym_help._is_value(shape):
+        shape = g.op("Constant", value_t=torch.LongTensor(shape))
+    return g.op("Reshape", self, shape, allowzero_i=0)
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset14.py:48]: i think for converting reshape from pytorch, the allowzero should be set to 1. Because pytorch reshape does not have semantic of "0 in shape preserving original value", 0 literally means 0. However, the blocker is that other symbolic functions may call `reshape` with 0 in shape and expect ONNX semantic. Please check if that's the case, and think of a proper design.

And in addition, may need to revise all the call sites in symbolic functions in other opsets, to invoke `reshape` based on actual opsets.

File: test/onnx/test_utility_funs.py [hunk 6]
```diff
@@ -731,7 +730,7 @@ def forward(self, x):
         assert next(iter).kind() == "aten::dequantize"
 
     # prim::ListConstruct is exported as onnx::SequenceConstruct for opset >= 11
-    @skipIfUnsupportedOpsetVersion([11, 12, 13])
+    @skipIfUnsupportedOpsetVersion([11, 12, 13, 14])
```
Comment (hwangdeyu) [inline test/onnx/test_utility_funs.py:733]: If this test is not support opset version >= 11, it may better to use skipIfUnsupportedMaxOpsetVersion to limit scope. Then we don't need to modify in the next opset version.

File: torch/onnx/symbolic_opset14.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+@parse_args("v", "v", "v", "v", "v", "i", "f", "f", "i")
+def batch_norm(g, input, weight, bias, running_mean, running_var, training, momentum, eps, cudnn_enabled):
```
Comment (jiafatom) [inline torch/onnx/symbolic_opset14.py:51]: Many part of the code seems similar to the implementation in symbolic_opset9, can we extract the common part and reuse the code?

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 2 above)
Comment (shubhambhokare1) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:5612]: Added :)

File: torch/onnx/symbolic_opset14.py (hunk 7 above)
Comment (shubhambhokare1) [reply torch/onnx/symbolic_opset14.py:51]: Good point, will wait on Negin's batchNorm PR #52758 for merge, to refactor the symbolic

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 8]
```diff
@@ -7571,6 +7627,7 @@ def forward(self, input):
         x = torch.randn(6, 4, 3, 3)
         self.run_test(FakeQuantizePerChannelModel(), (x))
 
+    @skipIfUnsupportedOpsetVersion([7, 8, 9, 10, 14])
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:7630]: nit: comment reason for the temporary skip.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 9]
```diff
@@ -596,16 +603,20 @@ c10::optional<std::vector<int64_t>> GetValueFromListConstructNode(
...
 
-void ProcessReshapeNode(Node* n) {
+void ProcessReshapeNode(Node* n, int opset_version) {
   if (ConstantValueMap::HasValue(n->input(1)->debugName())) {
     auto shape_temp =
         ConstantValueMap::GetValueInto1DInt64Vector(n->input(1)->debugName());
     auto shape_vector_0 =
         ConstantValueMap::GetShapeInto1DInt64VectorWithOneUnknown(
             n->input(0)->debugName());
+    int allowzero = 1;
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:613]: for older opsets the behavior in onnx should be allowzero = 0 right?

File: torch/onnx/symbolic_helper.py [hunk 10]
```diff
@@ -712,6 +712,16 @@ def _index_fill_reshape_helper(g, self, dim, index):
     expanded_index = expand(g, unsqueezed_index, expanded_index_shape, None)
     return expanded_index_shape, expanded_index
 
+# When using reshape helper (opset_version >= 14), if reshape has -1,
```
Comment (BowenBao) [inline torch/onnx/symbolic_helper.py:715]: Will this be relaxed to not allowing both -1 and 0 appearing together in reshape after Ashwini's fix?

File: torch/onnx/symbolic_opset14.py (hunk 7 above)
Comment (BowenBao) [reply torch/onnx/symbolic_opset14.py:41]: is this PR still blocked on #52758?

File: torch/csrc/jit/passes/onnx/constant_fold.h [hunk 11]
```diff
@@ -11,6 +11,7 @@ const int ONNX_OPSET_10 = 10;
 const int ONNX_OPSET_11 = 11;
 const int ONNX_OPSET_12 = 12;
 const int ONNX_OPSET_13 = 13;
+const int ONNX_OPSET_14 = 14;
```
Comment (fatcat-z) [inline torch/csrc/jit/passes/onnx/constant_fold.h:14]: Do we use this value somewhere? I guess we don't nee them any more since we changed the if statement in constant_fold.cpp file.

File: torch/onnx/symbolic_opset11.py [hunk 12]
```diff
@@ -177,7 +177,7 @@ def index_put(g, self, indices_list_value, values, accumulate=False):
     rank = sym_help._get_tensor_rank(values)
     if rank is not None and rank == 0:
         values = expand(g, values, values_shape, None)
-    values = g.op("Reshape", values, values_shape)
+    values = sym_help._reshape_helper(g, values, values_shape, allowzero=0)
```
Comment (fatcat-z) [inline torch/onnx/symbolic_opset11.py:180]: nit: Should we add an underscore in the middle(allowzero -> allow_zero) to follow the same naming rule as others?

File: torch/csrc/jit/passes/onnx/constant_fold.h (hunk 11 above)
Comment (shubhambhokare1) [reply torch/csrc/jit/passes/onnx/constant_fold.h:14]: Although we might not use it currently, I think it might be a good idea to leave it in there for future use

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp (hunk 9 above)
Comment (shubhambhokare1) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:613]: I don't this should affect shape inference as such, as the check is only valid for opset 14 onwards. I was wondering what would be the behavior if
1) node has no allowero
2) opset >= 14

File: torch/onnx/symbolic_helper.py (hunk 10 above)
Comment (shubhambhokare1) [reply torch/onnx/symbolic_helper.py:715]: Yes essentially, once Ashwinis fix is in, I will relax this condition and get rid of the workaround

File: torch/onnx/symbolic_opset11.py (hunk 12 above)
Comment (shubhambhokare1) [reply torch/onnx/symbolic_opset11.py:180]: I think allowzero should be fine as allowzero is the attr name used by ONNX spec. Also I was wondering which naming rule you were referring to?

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp (hunk 9 above)
Comment (jiafatom) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:613]: For the case you mentioned node has non allowzero and opset>=14, it should be same behavior as old opset, right?

Also I agree that `allowzero = 0` since this is the default value and back compatibility behavior for old opsets. It is true it does not affect shape inference, but seems more reasonable to set it to 0 by initialization.

File: torch/onnx/symbolic_helper.py [hunk 13]
```diff
@@ -712,6 +712,16 @@ def _index_fill_reshape_helper(g, self, dim, index):
...
+# allowzero cannot be set to 1
+def _reshape_helper(g, input, shape, allowzero=1):
```
Comment (jiafatom) [inline torch/onnx/symbolic_helper.py:717]: Seems it is more reasonable to set `allowzero=0` by default, because this is the default value for Reshape opset 14. Then we don't need have `allowzero=0` in any g.op("Reshape") call in old opsets.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp (hunk 9 above)
Comment (jiafatom) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:613]: nit: Can we change `ComputeShapeFromReshape` API to pass `Node* n` as argument? In that case, we don't need initialize `allow_zero` here, but just do this `opset_version >= 14` case inside `ComputeShapeFromReshape`. Then we only have one place to handle   `opset_version >= 14` .

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 8 above)
Comment (shubhambhokare1) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:7629]: Added reason for temporary skip

File: torch/onnx/symbolic_helper.py (hunk 13 above)
Comment (shubhambhokare1) [reply torch/onnx/symbolic_helper.py:717]: The reason allowzero=1 by default is because pytorch default behavior is reshape with allowzero=1

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp (hunk 9 above)
Comment (shubhambhokare1) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:631]: Updated this section

File: torch/onnx/symbolic_opset9.py [hunk 14]
```diff
@@ -70,11 +70,11 @@ def _shape_as_tensor(g, input):
 def _reshape_from_tensor(g, input, shape):
```
[the remark this answers is not in the document]
Comment (shubhambhokare1) [reply torch/onnx/symbolic_opset9.py:70]: Not sure if we should remove this as it is an at:: style op
https://pytorch.org/cppdocs/api/function_namespaceat_1a3d1d50cc33500dbf2a323bace7f8abe2.html

[the remark this answers is not in the document]
Comment (neginraoof) [reply torch/onnx/symbolic_opset9.py:70]: Got it, thanks. Assumed it is a helper.

File: torch/onnx/__init__.py [hunk 15]
```diff
@@ -13,7 +13,7 @@
 ir_version = _C._onnx.IR_VERSION
 producer_name = "pytorch"
 producer_version = _C._onnx.PRODUCER_VERSION
-constant_folding_opset_versions = [9, 10, 11, 12, 13]
+constant_folding_opset_versions = [9, 10, 11, 12, 13, 14]
```
[the remark this answers is not in the document]
Comment (shubhambhokare1) [reply torch/onnx/__init__.py:16]: Updated for reshape. Thanks

File: test/onnx/test_utility_funs.py [hunk 16]
```diff
@@ -917,6 +919,11 @@ def forward(self, x, y):
                                (TestCase,),
                                dict(TestUtilityFuns.__dict__, opset_version=13))
 
+# opset 14 tests
+TestUtilityFuns_opset14 = type(str("TestUtilityFuns_opset14"),
+                               (TestCase,),
+                               dict(TestUtilityFuns.__dict__, opset_version=14))
+
 # opset 11 tests
 TestUtilityFuns_opset11_new_jit_API = type(str("TestUtilityFuns_opset11_new_jit_API"),
```
Comment (BowenBao) [inline test/onnx/test_utility_funs.py:928]: nit: though unrelated, but seems these tests with name `new_jit_API` are no longer different from its above counterparts, can you remove them if that's the case?

**Review by BowenBao — APPROVED** [review]
LGTM, thanks!

File: test/onnx/test_models_onnxruntime.py [hunk 17]
```diff
@@ -7,7 +7,7 @@
 
 
 def exportTest(self, model, inputs, rtol=1e-2, atol=1e-7, opset_versions=None):
-    opset_versions = opset_versions if opset_versions else [7, 8, 9, 10, 11, 12]
+    opset_versions = opset_versions if opset_versions else [7, 8, 9, 10, 11, 12, 13]
```
[the remark this answers is not in the document]
Comment (neginraoof) [reply test/onnx/test_models_onnxruntime.py:10]: Please check.

**Review by neginraoof — CHANGES_REQUESTED** [review]

File: test/onnx/test_models_onnxruntime.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (shubhambhokare1) [reply test/onnx/test_models_onnxruntime.py]: Updated

**Review by neginraoof — APPROVED** [review]
Thanks!
